### ISSUE TITLE
too many read tcp  xxx> xxx i/o timeout

### ISSUE THREAD

**Issue by haochun** [denisenkom/go-mssqldb#331]
when i use sql server to save data.but,when the app run some day.there is too many  read tcp  xxx> xxx i/o timeout.does i use it error?

my code:

```
func getSQLConn(databaseUri string) (*sql.DB, error) {
	conn, ok := connections[databaseUri]
	if ok {
		return conn,nil
	}
	sqlLock.Lock()
	defer sqlLock.Unlock()
	conn, err := sql.Open("mssql", fmt.Sprintf(connectUri, databaseUri))
	if err != nil {
		return nil, err
	}
	conn.SetMaxOpenConns(config.EveryDBConnectionSize)
	conn.SetMaxIdleConns(config.EveryDBConnectionSize)
	connections[databaseUri] = conn
	return conn, nil
}

func insert(conn *sql.DB, action *ContentCommitAction) error {
	stmt, err := conn.Prepare(fmt.Sprintf(insertSQL, config.SqlDataDb, action.taskName))
	if err != nil {
		fmt.Printf("prepare stmt error,    info:%s\n", err.Error())
		return err
	}
	defer stmt.Close()

	_, err = stmt.Exec(action.content.Id, action.content.Url, action.content.Compress, action.content.Checksum, action.content.Data)
	if err != nil {
		if strings.Contains(err.Error(), "PRIMARY KEY") {
		} else {
			fmt.Printf("stmt exec error:%s    info:%s\n", action.taskName, err.Error())
			return err
		}
	}
	return nil
}
```

```
sqlConn, err := getSQLConn(databaseUri)
 if err != nil {
    fmt.Printf("task content commit service get sqlConnection error:,  info:%s\n", err)
    return
}
err = insert(sqlConn, action)
```
i nerver close the sql.DB

**Comment by haochun** [comment]
stmt exec error:china_com    info:read tcp 192.168.2.248:59110->192.168.2.248:1433: i/o timeout
stmt exec error:hexun_com    info:read tcp 192.168.2.248:58863->192.168.2.248:1433: i/o timeout
stmt exec error:hebnews_cn    info:read tcp 192.168.2.248:59122->192.168.2.248:1433: i/o timeout
stmt exec error:k618_cn    info:read tcp 192.168.2.248:59123->192.168.2.248:1433: i/o timeout
stmt exec error:yangtse_com    info:read tcp 192.168.2.248:59058->192.168.2.248:1433: i/o timeout

**Comment by haochun** [comment]
@vadimyer did you resolve this problem?

**Comment by vadimyer** [comment]
No, I currently do not know what causes this problem. But I guess this should be handled gracefully by the driver probably. Though, if it's a real timeout problem, the driver can't do much about it, except some retry mechanism.

**Comment by baodinh** [comment]
i have same problem too.

**Comment by paracha3** [comment]
Same problems here too. Funny thing is same pieces of compiled code would give this error the first time, you run the binary again and it runs flawlessly. I have no idea what causes this error and when and how to avoid it.

**Comment by m82labs** [comment]
The only time I see this error is if I am running a query that takes longer than 30 seconds to execute, and I can't seem to find a way to adjust that timeout. I can change the connection timeout but not the timeout for actual query execution. Is this the issue you are experiencing?

**Comment by paracha3** [comment]
@m82labs yes same here. Funny i see it the first time i run code because SQL server takes long and fails. I run the same binary again and my guess is that SQL Server has cached some stuff so it returns faster and i do not get this error.

**Comment by m82labs** [comment]
@paracha3 is you just execute the query 
```WAITFOR DELAY '00:00:40'```
It will time out every time.

Other drivers (like the mysql driver) seem to be able to handle a "command timeout" setting. I would love to see the same for this driver. I am a novice at golang and can try to implement it myself, but it make take a little time.

**Comment by Zenlil** [comment]
I too have the same problem, got a query that needs to run and months end that takes between 2-3 minutes. Since I'm also using Azure SQL-service there is "jobs" to start in the background, and I do not want to resort to create a C#.Net-component just to be able to use the SqlCommand.CommandTimeout (https://msdn.microsoft.com/en-us/library/system.data.sqlclient.sqlcommand.commandtimeout(v=vs.110).aspx)
So, please add the option of controlling the 'CommandTimeout' or instruct us on how to do it because we currently can't find a solution in Go and MSSQL

**Comment by kardianos** [comment]
I don't think this is a command timeout issue. I think this is a bug. I haven't looked into it yet.

**Comment by kardianos** [comment]
This may be a connection timeout issue. Can you try to get latest from the driver and set the following on the DSN connection string: `connection timeout=0` ?

**Comment by Zenlil** [comment]
**main.go**
```
package main

import (
	"database/sql"
	"log"
	"time"

	_ "github.com/denisenkom/go-mssqldb"
)

const DSN = "server=XXXX;database=YYYY;User ID=ZZZZ;Password=####

"

func main() {
	db, err := sql.Open("mssql", DSN)
	if err != nil {
		panic(err)
	}

	log.Println("call Exec Waitfor 28 seconds")
	t := time.Now()
	_, err = db.Exec("WAITFOR DELAY '00:00:28'")
	if err != nil {
		log.Println("Error after", time.Since(t))
		panic(err)
	}
	log.Println("done waitfor 28 seconds - duration", time.Since(t))

	log.Println("call Exec Waitfor 32 seconds")
	t = time.Now()
	_, err = db.Exec("WAITFOR DELAY '00:00:32'")
	if err != nil {
		log.Println("Error after", time.Since(t))
		panic(err)
	}
	log.Println("done waitfor 32 seconds - duration", time.Since(t))

	db.Close()
}
```


```
$ go version
 go version go1.9.2 linux/amd64
```

```
$ go get -u github.com/denisenkom/go-mssqldb
```

```
$ go run main.go
2018/03/16 09:20:25 call Exec Waitfor 28 seconds
2018/03/16 09:20:53 done waitfor 28 seconds - duration 28.370434s
2018/03/16 09:20:53 call Exec Waitfor 32 seconds
2018/03/16 09:21:23 Error after 30.047447s
panic: read tcp 10.101.102.56:61884->#.#.#.#:1433: i/o timeout

goroutine 1 [running]:
main.main()
        main.go:33 +0x604
exit status 2
```

Result: This is not a connection-problem, this is a command timeout problem, sadly since the command timeout can't be set using the DSN.

**Comment by kardianos** [comment]
Try setting the connection timeout to zero 0 in the dsn with latest driver version.

**Comment by Zenlil** [comment]
I added `;Connection Timeout=0` to the end of the DSN I posted above and got the following result:
```
$ go run main.go
2018/03/16 13:49:00 call Exec Waitfor 28 seconds
2018/03/16 13:49:28 done waitfor 28 seconds - duration 28.566814s
2018/03/16 13:49:28 call Exec Waitfor 32 seconds
2018/03/16 13:50:00 done waitfor 32 seconds - duration 32.061654s
```

This is good news to have as a workaround, but it does not match how a default Microsoft SQL Server behaves.
See:
- https://www.w3schools.com/asp/prop_commandtimeout.asp
- https://www.connectionstrings.com/all-sql-server-connection-string-keywords/

I do understand that this Go-package doesn't have a corresponding SqlCommand  as .NET or ADO has, but I do not fancy re-using an already existing property on Connection timeouts to be re-purposed into something else.
The option of having separate timeouts per 'Command' is much more appealing as we then can choose a short timeout on things supposed to go quickly, and explicit longer execution-time on the few commands that we actually want to allow to spend more time.

Please do not think of me as an angry user, I actually find this package just awesome. But my main focus is the consistent work and usage of databases with expected behaviors, and re-purposing the Connection timeout into a dual meaning like this is the 'dark side' i.m.o 

Since this package lacks the SqlCommand my suggestion is this:
Add a function on the Prepared statement to allow settings an explicit 'Command Timeout'
That way the default 30 seconds is valid for all default 'commands' and if one need to enable long-running queries then just make them do 'Prepare' and set an explicit value.

**Comment by kardianos** [comment]
I will look into the consistently issue.

Go has a vastly more flexible solution. Use context for per command timeout.

**Comment by Lercher** [comment]
Just in case s/o (like I did) stumbles into this dicussion: having the said connection timeout value in the connection string, it takes precedence over any context.Context that is handed over to e.g. QueryContext. I.e. even using context.Background() explicitly, a query times out after the (positive) connection timeout (checked today).

Wouldn't it be better to ignore the connection string value, if a context with non-infinite deadline is given? Is that a breaking change or a bug fix?
